test/distributed/test_ddp_with_rpc.py

```diff
@@ -0,0 +1,541 @@
+        )
+        # The same size of mini batch.
+        assert sparse.shape[0] == input.dense_features.shape[0]
+        dense = self.fc1(input.dense_features)
+        x = torch.cat((dense, sparse), 1)
+        gLogger.debug(f"Concatenated feature: {x}")
+        x = _remote_method(RemoteNet.forward, self.remote_net_rref, x)
+        return self.fc2(x)
+
+    def print_parameters(self):
+        gLogger.debug(f"Local parameters: {list(self.parameters())}")
+        gLogger.debug(
+            f"Remote parameters: {[param_rref.to_here() for param_rref in self.remote_parameters_rrefs]}"
+        )
+
+
+class Trainer:
+    def __init__(
+        self,
+        remote_em_rref: rpc.RRef,
+        remote_net_rref: rpc.RRef,
+        ddp_mode: DdpMode,
+        rank: int,
+    ):
+        self.process_group_context = ProcessGroupContext(
+            name=TRAINER_NAME_TEMPLATE.format(rank),
+            ranks=DDP_TRAINER_RANKS,
+            group_name="trainers_ddp",
+        )
+        process_group_for_ddp = self.process_group_context.__enter__()
+        check_process_group_type(f"trainer #{rank}", process_group_for_ddp)
+        gLogger.info(
+            f"Initing a trainer with process group {process_group_for_ddp} ..."
+        )
+
+        self.criterion = torch.nn.CrossEntropyLoss()
+        self.remote_em_rref = remote_em_rref
+        self.remote_net_rref = remote_net_rref
+        self.hybrid_module = HybridModel(
+            self.remote_em_rref,
+            self.remote_net_rref,
+            process_group_for_ddp if ddp_mode in (DdpMode.INSIDE,) else None,
+        )
+        if ddp_mode == DdpMode.OUTSIDE:
+            gLogger.info(f"Wrapping the whole hybride module into DDP.")
+            self.hybrid_module = DistributedDataParallel(
+                self.hybrid_module,
+                process_group=process_group_for_ddp,
+                check_reduction=True,
+            )
+        self.local_optimizer = optim.SGD(self.hybrid_module.parameters(), lr=LR)
+        gLogger.info(f"Succeeded in creating a HybridModel instance.")
+
+    def __del__(self):
+        self.process_group_context.__exit__(None, None, None)
+
+    def do_forward_without_grad(self, input: FeatureSet):
+        gLogger.debug(f"Doing a forward pass on {input}")
+        with torch.no_grad():
+            output = self.hybrid_module(input)
+            return self.criterion(output, input.labels)
+
+    def get_real_module(self):
+        if isinstance(self.hybrid_module, DistributedDataParallel):
+            return self.hybrid_module.module
+        else:
+            return self.hybrid_module
+
+    def print_parameters(self):
+        self.get_real_module().print_parameters()
+
+    def do_mini_batch(self, mini_batch: FeatureSet):
+        gLogger.info(f"Doing a mini batch on {mini_batch}")
+        loss = 0
+
+        with dist_autograd.context() as context_id:
+            self.local_optimizer.zero_grad()
```

Comment:
We should use only one optimizer in the unit test (DistributedOptimizer)

Reply:
I'll delete it. Just confirmed it doesn't work at all.